src/huggingface_hub/serialization/_base.py

```diff
@@ -0,0 +1,163 @@
+        get_storage_id (`Callable[[Tensor], Optional[Any]]`, *optional*):
+            A function that returns a unique identifier to a tensor storage. Multiple different tensors can share the
+            same underlying storage. This identifier is guaranteed to be unique and constant for this tensor's storage
+            during its lifetime. Two tensor storages with non-overlapping lifetimes may have the same id.
+
+    Returns:
+        [`StateDictSplit`]: A `StateDictSplit` object containing the shards and the index to retrieve them.
+    """
+    storage_id_to_tensors: Dict[Any, List[str]] = {}
+
+    shard_list: List[Dict[str, TensorT]] = []
+    current_shard: Dict[str, TensorT] = {}
+    current_shard_size = 0
+    total_size = 0
+
+    for key, tensor in state_dict.items():
+        # when bnb serialization is used the weights in the state dict can be strings
+        # check: https://github.com/huggingface/transformers/pull/24416 for more details
+        if isinstance(tensor, str):
+            continue
+
+        # If a `tensor` shares the same underlying storage as another tensor, we put `tensor` in the same `block`
+        storage_id = get_storage_id(tensor)
+        if storage_id is not None:
+            if storage_id in storage_id_to_tensors:
+                # We skip this tensor for now and will reassign to correct shard later
+                storage_id_to_tensors[storage_id].append(key)
+                continue
+            else:
+                # This is the first tensor with this storage_id, we create a new entry
+                # in the storage_id_to_tensors dict => we will assign the shard id later
+                storage_id_to_tensors[storage_id] = [key]
+
+        # Compute tensor size
+        tensor_size = get_tensor_size(tensor)
+
+        # If this tensor is bigger than the maximal size, we put it in its own shard
+        if tensor_size > max_shard_size:
+            total_size += tensor_size
+            shard_list.append({key: tensor})
+            continue
+
+        # If this tensor is going to tip up over the maximal size, we split.
+        # Current shard already has some tensors, we add it to the list of shards and create a new one.
+        if current_shard_size + tensor_size > max_shard_size:
+            shard_list.append(current_shard)
+            current_shard = {}
+            current_shard_size = 0
+
+        # Add the tensor to the current shard
+        current_shard[key] = tensor
+        current_shard_size += tensor_size
+        total_size += tensor_size
+
+    # Add the last shard
+    if len(current_shard) > 0:
+        shard_list.append(current_shard)
+    nb_shards = len(shard_list)
+
+    # Loop over the tensors that share the same storage and assign them together
+    for storage_id, keys in storage_id_to_tensors.items():
+        # Let's try to find the shard where the first tensor of this storage is and put all tensors in the same shard
+        for shard in shard_list:
```

Comment:
Alternative to the loop: do the assignment currently in L103 after the shard has been assigned for the first tensor in the current storage_id, and put there the shard as well. Probably more confusing.

Reply:
Yeah in the end I went for the solution "build a list of pairs and assign once all tensors have been assigned to a shard" because tensors can be assigned to shards in two ways: 1. by being assigned to the "current shard" and 2. by being added to the shard_list as a standalone shard (if too big). As a result, it's harder to know in which shard id the tensor will be. For what I have tested, I think this results in a bug in `transformers` when we compute the [`block_id`](https://github.com/younesbelkada/transformers/blob/d812cf29d87a02737dd0a84cf39c23eaf6f31af8/src/transformers/modeling_utils.py#L317). If the "next tensor" is bigger than the shard size, it takes the place immediately which shifts the `block_id`. In practice, I believe it never happens because shard size is much larger than each single tensors.

All of this to say, I prefer the current implementation where we explicitly assign the tensors sharing the same storage_id _after_ all "normal" tensors have been added to a shard. Worse case scenario, this is an extra O(N*M) operation with N files and M shards => not very problematic IMO.